core/src/main/scala/kafka/log/LogValidator.scala

```diff
@@ -102,36 +111,44 @@ private[kafka] object LogValidator {
 
   private def assignOffsetsNonCompressed(records: MemoryRecords,
                                          offsetCounter: LongRef,
-                                         now: Long,
+                                         currentTimestamp: Long,
                                          compactedTopic: Boolean,
                                          timestampType: TimestampType,
                                          timestampDiffMaxMs: Long): ValidationAndOffsetAssignResult = {
-    var maxTimestamp = Record.NO_TIMESTAMP
+    var maxTimestamp = RecordBatch.NO_TIMESTAMP
     var offsetOfMaxTimestamp = -1L
-    val firstOffset = offsetCounter.value
+    val initialOffset = offsetCounter.value
 
-    for (entry <- records.shallowEntries.asScala) {
-      val record = entry.record
-      validateKey(record, compactedTopic)
+    for (batch <- records.batches.asScala) {
+      ensureNonTransactional(batch)
 
-      val offset = offsetCounter.getAndIncrement()
-      entry.setOffset(offset)
+      for (record <- batch.asScala) {
+        record.ensureValid()
+        ensureNotControlRecord(record)
+        validateKey(record, compactedTopic)
 
-      if (record.magic > Record.MAGIC_VALUE_V0) {
-        validateTimestamp(record, now, timestampType, timestampDiffMaxMs)
+        val offset = offsetCounter.getAndIncrement()
+        if (batch.magic > RecordBatch.MAGIC_VALUE_V0) {
+          validateTimestamp(batch, record, currentTimestamp, timestampType, timestampDiffMaxMs)
 
-        if (timestampType == TimestampType.LOG_APPEND_TIME)
-          entry.setLogAppendTime(now)
-        else if (record.timestamp > maxTimestamp) {
-          maxTimestamp = record.timestamp
-          offsetOfMaxTimestamp = offset
+          if (record.timestamp > maxTimestamp) {
+            maxTimestamp = record.timestamp
+            offsetOfMaxTimestamp = offset
+          }
         }
       }
+
+      batch.setLastOffset(offsetCounter.value - 1)
+
+      // TODO: in the compressed path, we ensure that the batch max timestamp is correct.
+      //       We should either do the same or (better) let those two paths converge.
```

Comment:
Should we file a JIRA for this? We generally avoid leaving TODOs in the code (this comment is useful, so maybe we just remove the TODO marker).